src/fu-agent.c

```diff
@@ -0,0 +1,645 @@
+	gboolean verbose = FALSE;
+	g_autoptr(FuUtilPrivate) priv = g_new0 (FuUtilPrivate, 1);
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GPtrArray) cmd_array = fu_util_cmd_array_new ();
+	g_autofree gchar *cmd_descriptions = NULL;
+	const GOptionEntry options[] = {
+		{ "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose,
+			/* TRANSLATORS: command line option */
+			_("Show extra debugging information"), NULL },
+		{ NULL}
+	};
+
+	setlocale (LC_ALL, "");
+
+	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
+	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
+	textdomain (GETTEXT_PACKAGE);
+
+	/* ensure D-Bus errors are registered */
+	fwupd_error_quark ();
+
+	/* create helper object */
+	priv->loop = g_main_loop_new (NULL, FALSE);
+
+	/* add commands */
+	fu_util_cmd_array_add (cmd_array,
+			       "sync", NULL,
+			       /* TRANSLATORS: command description */
+			       _("Sync the current system status with the server"),
+			       fu_util_sync);
+	fu_util_cmd_array_add (cmd_array,
+			       "register", NULL,
+			       /* TRANSLATORS: command description */
+			       _("Register with a remote server"),
+			       fu_util_register);
+	fu_util_cmd_array_add (cmd_array,
+			       "unregister", NULL,
+			       /* TRANSLATORS: command description */
+			       _("Unregister with a remote server"),
+			       fu_util_unregister);
+	fu_util_cmd_array_add (cmd_array,
+			       "enable", "SERVER",
+			       /* TRANSLATORS: command description */
+			       _("Enable by configuring a server to use"),
+			       fu_util_enable);
+	fu_util_cmd_array_add (cmd_array,
+			       "disable", NULL,
+			       /* TRANSLATORS: command description */
+			       _("Disable by clearing the configured server"),
+			       fu_util_disable);
+
+	/* sort by command name */
+	fu_util_cmd_array_sort (cmd_array);
+
+	/* do stuff on ctrl+c */
+	priv->cancellable = g_cancellable_new ();
+	g_unix_signal_add_full (G_PRIORITY_DEFAULT,
+				SIGINT, fu_util_sigint_cb,
+				priv, NULL);
+
+	/* get a list of the commands */
+	priv->context = g_option_context_new (NULL);
+	cmd_descriptions = fu_util_cmd_array_to_string (cmd_array);
+	g_option_context_set_summary (priv->context, cmd_descriptions);
+	g_option_context_set_description (priv->context,
+		"This tool reports the current firmware status back to a central server.");
+
+	/* TRANSLATORS: program name */
+	g_set_application_name (_("Firmware Agent"));
+	g_option_context_add_main_entries (priv->context, options, NULL);
+	ret = g_option_context_parse (priv->context, &argc, &argv, &error);
+	if (!ret) {
+		/* TRANSLATORS: the user didn't read the man page */
+		g_print ("%s: %s\n", _("Failed to parse arguments"),
+			 error->message);
+		return EXIT_FAILURE;
+	}
+
+	/* set verbose? */
+	if (verbose) {
+		g_setenv ("G_MESSAGES_DEBUG", "all", FALSE);
+	} else {
+		g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG,
+				   fu_util_ignore_cb, NULL);
+	}
+
+	/* connect to the daemon */
+	priv->client = fwupd_client_new ();
```

Comment:
connecting to daemon this early means that if no arguments are provided (or an invalid command is provided) then it still connects to the daemon.  This means spinning up the daemon needlessly.  I'd say move this into each function instead after argument validation.

Reply:
Nope; just creating the object doesn't do a daemon connection. Any `fwupd_client_*()` method magically calls `fwupd_client_connect()` automatically. Another clue is that `fwupd_client_new()` is not _failable_, i.e. can't return a `GError`.

Reply:
Sorry I'm a line off on the comment, but the connection is just one line lower and that's what I am saying should move.

Reply:
Then I apologize for my arrogance. :)